include/nbl/video/utilities/CComputeBlit.h

```diff
@@ -0,0 +1,828 @@
+	{
+		if (scratchBuffer)
+		{
+			outBufferRange->offset = 0ull;
+			outBufferRange->size = core::alignUp(sizeof(uint32_t), logicalDevice->getPhysicalDevice()->getLimits().SSBOAlignment);
+			outBufferRange->buffer = scratchBuffer;
+		}
+	}
+
+	static inline void getDefaultAlphaHistogramBufferRange(asset::SBufferRange<video::IGPUBuffer>* outBufferRange, core::smart_refctd_ptr<video::IGPUBuffer> scratchBuffer, video::ILogicalDevice* logicalDevice)
+	{
+		if (scratchBuffer)
+		{
+			outBufferRange->offset = core::alignUp(sizeof(uint32_t), logicalDevice->getPhysicalDevice()->getLimits().SSBOAlignment);
+			outBufferRange->size = sizeof(uint32_t) * DefaultAlphaBinCount;
+			outBufferRange->buffer = scratchBuffer;
+		}
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> getDefaultAlphaTestDSLayout()
+	{
+		if (!alphaTestDSLayout)
+		{
+			constexpr uint32_t DESCRIPTOR_COUNT = 2u;
+			asset::E_DESCRIPTOR_TYPE types[DESCRIPTOR_COUNT] = { asset::EDT_COMBINED_IMAGE_SAMPLER, asset::EDT_STORAGE_BUFFER }; // input image, alpha test atomic counter
+			alphaTestDSLayout = getDSLayout(DESCRIPTOR_COUNT, types, device, sampler);
+		}
+
+		return alphaTestDSLayout;
+	}
+
+	inline core::smart_refctd_ptr<video::IGPUPipelineLayout> getDefaultAlphaTestPipelineLayout()
+	{
+		if (!alphaTestPipelineLayout)
+		{
+			asset::SPushConstantRange pcRange = {};
+			{
+				pcRange.stageFlags = asset::IShader::ESS_COMPUTE;
+				pcRange.offset = 0u;
+				pcRange.size = sizeof(alpha_test_push_constants_t);
+			}
+
+			alphaTestPipelineLayout = device->createGPUPipelineLayout(&pcRange, &pcRange + 1ull, core::smart_refctd_ptr(alphaTestDSLayout));
+		}
+
+		return alphaTestPipelineLayout;
+	}
+
+	core::smart_refctd_ptr<video::IGPUSpecializedShader> createAlphaTestSpecializedShader(const asset::IImage::E_TYPE inImageType)
+	{
+		const char* shaderpath = "../default_compute_alpha_test.comp";
+		auto cpuShader = loadShaderFromFile(shaderpath);
+
+		if (!cpuShader)
+			return nullptr;
+
+		const auto workgroupDims = getDefaultWorkgroupDims(inImageType);
+
+		auto cpuShaderOverriden = asset::IGLSLCompiler::createOverridenCopy(cpuShader.get(),
+			"#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d\n"
+			"#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d\n"
+			"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
+			"#define _NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ %d\n",
+			inImageType >= asset::IImage::ET_1D ? workgroupDims.x : 1u,
+			inImageType >= asset::IImage::ET_2D ? workgroupDims.y : 1u,
+			inImageType >= asset::IImage::ET_3D ? workgroupDims.z : 1u,
+			static_cast<uint32_t>(inImageType) + 1u);
+
+		cpuShaderOverriden->setFilePathHint(shaderpath);
+
+		auto gpuUnspecShader = device->createGPUShader(std::move(cpuShaderOverriden));
+
+		return device->createGPUSpecializedShader(gpuUnspecShader.get(), { nullptr, nullptr, "main" });
```

Comment:
its best just to create a small `ICPUShader` in memory from an `ICPUBuffer` which is a view over a string, then give that  to the compiler (and rely on it to get all the other code via include)

its like 5 lines and you don't need a `loadShaderFromFile`

for example you'd do
```cpp
const char* shaderStart = "#include \"nbl/builtin/glsl/blit/default_compute_alpha_test.comp\"";
auto shader = core::make_smart_refctd_ptr<CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(strlen(shaderStart),shaderStart,core::adopt_memory) ;

auto cpuShader = core::make_smart_refctd_ptr<ICPUShader>(std::move(shader),ICPUShader::buffer_contains_glsl,ESS_COMPUTE,"CComputeBlit::createAlphaTestSpecializedShader");
auto cpuShaderOverriden = asset::IGLSLCompiler::createOverridenCopy(cpuShader.get(),
			"#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d\n"
			"#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d\n"
			"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
			"#define _NBL_GLSL_BLIT_ALPHA_TEST_DIM_COUNT_ %d\n",
			inImageType >= asset::IImage::ET_1D ? workgroupDims.x : 1u,
			inImageType >= asset::IImage::ET_2D ? workgroupDims.y : 1u,
			inImageType >= asset::IImage::ET_3D ? workgroupDims.z : 1u,
			static_cast<uint32_t>(inImageType) + 1u);
```

Reply:
Resolved.